src/org/exist/storage/NativeBroker.java

```diff
@@ -2532,44 +2483,117 @@ public void copyResource(final Txn transaction, final DocumentImpl doc, final Co
                     trigger.afterDeleteDocument(this, transaction, newURI);
                 }
 
-                trigger.beforeCopyDocument(this, transaction, doc, newURI);
+                doCopyDocument(transaction, trigger, doc, destination, newName);
+
+            } catch(final ReadOnlyException e) {
+                throw new IOException(e.getMessage(), e);
+            } catch(final TriggerException e) {
+                throw new PermissionDeniedException(e.getMessage(), e);
+            } finally {
+                lock.release(LockMode.WRITE_LOCK);
+            }
+        }
+    }
+
+    /**
+     * Creates a new Document object for the destination document
+     * - copies the nodes from the source document to the destination document
+     * - if no existing document in the destination:
+     *      - adds the destination document to the destination collection
+     *   else, switches the existing document object for the new document in the destination collection
+     *
+     *   asynchronously deletes the nodes of the old existing document
+     */
+    private void doCopyDocument(final Txn transaction, final DocumentTrigger trigger,
+                                final DocumentImpl srcDoc, final Collection dstCol, final XmldbURI dstDocName)
+            throws TriggerException, LockException, PermissionDeniedException, IOException, ReadOnlyException, EXistException {
+
+        // URI of the destination document in the database
+        final XmldbURI dstDocUri = dstCol.getURI().append(dstDocName);
+
+        trigger.beforeCopyDocument(this, transaction, srcDoc, dstDocUri);
+
+        final DocumentImpl currentDstDoc = dstCol.getDocumentWithLock(this, dstDocName, LockMode.READ_LOCK);
```

Comment:
can it simply call `delete` operation for destination document? Mean, before start copy operation.

Reply:
That would be more expensive.

Reply:
it'll be the proper way! you should not only index inform about document remove but also triggers.

Reply:
@shabanovd I am not sure I follow you. Can you be clearer please? I have not changed how the triggers work here.

Reply:
and that is the problem. When overwriting resource it must be deleted by "remove" procedure to inform index and fire trigger events.

Reply:
@shabanovd My goal in this PR, was not to change on which events triggers were fired for from how they were before this PR. If you would like to change that, could you submit another PR once this one is merged please?